earthpy/spatial.py

```diff
@@ -370,6 +370,107 @@ def crop_image(raster, geoms, all_touched=True):
     return out_image, out_meta
 
 
+def crop_all(
+    band_paths,
+    output_dir,
+    geoms,
+    overwrite=False,
+    all_touched=True,
+    verbose=True,
+):
+    """Takes a list of rasters and a boundary, and crops them efficiently.
+
+    Parameters
+    ----------
+    band_paths : list of file paths
+        List of paths of bands that are to be cropped.
+    output_dir : list or string
+        If a list is provided, then the output files from this function will be written to
+        each path in the list. This allows for customization of file names.
+        If a string is provided, then the output files from the function will be written
+        to that directory with the name of the original file plus a suffix of "_crop".
+    geoms : geopandas geodataframe or list of polygons
+        The spatial polygon boundaries in GeoJSON-like dict format
+        to be used to crop the image. All data outside of the polygon
+        boundaries will be set to nodata and/or removed from the image.
+    overwrite : bool (default=False)
+        Disallows files to be overwritten if they exist already. Can be changed so that files
+        can be overwritten with each run of the function.
+    all_touched : bool (default=True)
+        Include a pixel in the mask if it touches any of the
+        shapes. If False, include a pixel only if its center is within one of
+        the shapes, or if it is selected by Bresenham's line algorithm.
+        (from rasterio)
+    verbose : bool (default=True)
+        Returns a list of file paths with the files that were created in the function. Can
+        be turned off so the user doesn't have to assign the function to a variable.
+
+    Returns
+    ----------
+    return files : list
+        List of the files created by the function.
+
+    Example
+    -------
+        >>> import os
+        >>> import earthpy.spatial as es
+        >>> import geopandas as gpd
+        >>> from earthpy.io import path_to_example
+        >>> band_fnames = ["red.tif", "green.tif", "blue.tif"]
+        >>> band_paths = [path_to_example(fname) for fname in band_fnames]
+        >>> rmnp = gpd.read_file(path_to_example("rmnp.shp"))
+        >>> output_dir = os.path.commonpath(band_paths)
+        >>> output_files = es.crop_all(band_paths, output_dir, rmnp, overwrite=True)
+        >>> len(output_files)
+        3
+        >>> os.path.isfile(output_files[0])
+        True
+        >>> # Cleaning up example data
+        >>> for bands in output_files:
+        ...     os.remove(bands)
+
+    """
+    try:
+        if not os.path.exists(os.path.commonpath(output_dir)):
```

Comment:
@nkorinek why are you using os.path.commonpath here? i'm just unclear about what you are doing. if the user provides a path, that path should work properly.

Reply:
I'm using `os.path.comonpath()` here to get the output directory of all the rasters they provided if the user passes in a list of paths. The files don't exist yet, so I can't check those, but I can check to make sure that the common directory that the user is planning to store them in exists. Is there a different way I should be checking this?

Reply:
@nkorinek the logic here doesn't seem to work as i would expect it to. can you tell me - in this try statement you assume if the path exists that there is a list. however i just tried it with a list and it returned the root dirs for each file.

all of the statements below work for lists. and it doesn't seem to test whether the dirs all exist 

```
output_dir = ["/Users/leahwasser/Documents/test2/im.tif", "/Users/leahwasser/Documents/test/im2.tif"]
os.path.exists(os.path.commonpath(output_dir))
```

```
output_dir = ["/Users/leahwasser/Documents/test/im.tif", "/Users/leahwasser/Documents/test/im2.tif"]
os.path.exists(os.path.commonpath(output_dir))
```

We might need a few additional tests here but first... i'd like a bit of input on the dev / usability side. 

I think we might need a different approach. i'd like to ping @mbjoseph to see what he thinks.
Max -- essentially here we are allowing `output_dir` to be either a single path or a list of paths with a file name associated with it. From a user perspective, there are a few ways this is confusing to me

1. they can provide a list or a string. 
2. they can provide a directory only or a list of file names with an associated directory. 

how would you expect this to behave?

A few options.
1. we only allow lists and test for the length of the list. if the `len >1` then we can handle a list of file names (assuming they are there) 
2. have two arguments - the user must use one of them but `out_dir` could be a single directory, `out_files `or something could be a list of paths w associated file names. 

thoughts?

Reply:
@lwasser this might be the path to go, and I was thinking about something along the lines of option 2 while I was writing this function. I like both options though! you're right, some of the functionality is a bit weird. The issue is my tests assume that all of the files in a list are pointing to the same folder, which I shouldn't have assumed. Didn't even think about that! This does require changes. Another option is that we could loop through the list and test for each individual folder.

Reply:
ok max replied on slack. @nkorinek let's simplify things here. for the time being, let's only accept a single path for out_dir. then all images will be written to that path. this will simplify this part of the code considerably.